xdsl/dialects/func.py

```diff
@@ -69,6 +69,51 @@ def from_region(name: str, input_types: List[Attribute],
         op = FuncOp.build(attributes=attributes, regions=[region])
         return op
 
+    def replace_argument_type(self, arg: int | BlockArgument,
+                              new_type: Attribute):
+        """
+        Replaces a function argument with a new one
+        """
+
+        if isinstance(arg, int):
+            if len(self.body.blocks[0].args) <= arg:
+                raise IndexError("Block {} does not have argument #{}!".format(
+                    self.body.blocks[0], arg))
+            arg = self.body.blocks[0].args[arg]
+
+        arg.typ = new_type
+        self.update_function_type()
+
+    def update_function_type(self):
+        """
+        Update the function_type attribute to reflect changes in the
+        block argument types.
+        """
+        return_op = self.body.blocks[-1].ops[-1]
+        return_type: tuple[Attribute] = tuple()
+
+        if return_op is not None:
+            return_type = tuple(arg.typ for arg in return_op.operands)
+
+        self.attributes['function_type'] = FunctionType.from_lists(
+            [arg.typ for arg in self.args],
+            return_type,
+        )
+
+    def get_return_op(self) -> Return | None:
+        if len(self.body.blocks) == 0:
+            return
+        if len(self.body.blocks[-1].ops) == 0:
+            return
+        # TODO: remove once we have verify() check this!
+        if not isinstance(self.body.blocks[-1].ops[-1], Return):
+            return
+        return self.body.blocks[-1].ops[-1]
+
+    @property
+    def args(self) -> tuple[BlockArgument, ...]:
+        return self.body.blocks[0].args
```

Comment:
Hmm, here you don't check if the block exist, but you did previously.
I guess you have a choice here, either you can decide to return `None`, or you can decide to raise an Exception.

My feelings would be that you want to return an exception both here and in `get_return_op`, to remain consistent. Also, you probably never want to call `get_return_op` if there are no blocks.
MLIR has a `is_declaration` method to get that information directly.

Reply:
Added!